core/arch/arm/plat-rcar/romapi_call.S

```diff
@@ -0,0 +1,87 @@
+/* SPDX-License-Identifier: BSD-2-Clause */
+/*
+ * Copyright (c) 2021, EPAM Systems
+ */
+
+#include <asm.S>
+#include <arm.h>
+#include <arm64_macros.S>
+#include <kernel/cache_helpers.h>
+
+/* uint32_t __plat_call_romapi(paddr_t func, uint64_t arg1, uint64_t arg2,
+ *			       uint64_t arg3)
+ * Call MaskROM function func(arg1, arg2, arg3).
+ * We need to disable MMU before calling any MaskROM API functions
+ */
+FUNC __plat_call_romapi , : , .identity_map
+
+	push	fp,  lr
+	push	x19, x20
+	push	x21, x22
+	push	x23, x24
+
+	mov	x19, x0
+	mov	x20, x1
+	mov	x21, x2
+	mov	x22, x3
+
+	/* Get PA of stack pointer */
+	mov	x0, sp
+#ifdef CFG_CORE_ASLR
+	/*
+	 * We are running at identity location, so we can't use bl there,
+	 * becase assemlber will generate relative address to virt_to_phys(),
+	 * which is not identity mapped. So we need to use absolute address.
+	 */
+	ldr	x9, =virt_to_phys
+	blr	x9
+#else
+	bl	virt_to_phys
+#endif
+	mov	x23, x0
+
+	/* We about to disable MMU. Make sure that all writes reached memory */
+	mov     x0, #DCACHE_OP_CLEAN
+	bl	dcache_op_all
+
+	/* Disable MMU */
+	mrs	x9, sctlr_el1
+	and	x9, x9, #(~SCTLR_M)
+	and	x9, x9, #(~SCTLR_C)
```

Comment:
Why do you need to turn off the data cache? Some requirement by the ROM? If you were able to keep it on there's a good chance that you wouldn't need to do any cache maintenance at all.

Reply:
Hmmm, maybe my understanding of VMSA is incorrect. When MMU is disabled, all memory is treated as Device-nGnRnE, so no data cache can be used. I'm not sure how data cache will behave in this case. 

According to D4.10.3 "If the change affects the cacheability attributes of the area of memory, including any change between Write-Through and Write-Back attributes, software must ensure that any cached copies of affected locations are removed from the caches, typically by cleaning and invalidating the locations from the levels of cache that might hold copies of the locations affected by the attribute change". I'm assuming that disabling MMU can be considered as "changing cacheability attributes".

So, maybe it is not mandatory to clear C bit. But I need to do perform cache maintenance anyways. If my understanding of ARM ARM is correct, of course :)

Reply:
I believe this changed with Aarch64, but I don't have a reference right now.

Reply:
Well, practice is the criterion of truth :) OP-TEE panics if I'm not clearing `SCTLR_C`

But, now I'm seeing a new issue which I didn't saw earlier. I need to investigate it...